ENH: Stop using siemens2rads from old nipype workflows

oesteban opened the pull request:
A new interface is proposed, to be used in phase-difference based
workflows.

@AKSoo identified a potential issue with siemens2rads
(https://github.com/poldracklab/sdcflows/pull/30#discussion_r346640162),
although the re-centering did not seem to make a big difference.

This implementation of the rescaling uses percentiles to robustly
calculate the range of the input image (as recommended in the original
paper about FSL PRELUDE), and makes sure the output data type is
adequate (float32).

oesteban committed:
ENH: Stop using siemens2rads from old nipype workflows

A new interface is proposed, to be used in phase-difference based
workflows.

@AKSoo identified a potential issue with siemens2rads
(https://github.com/poldracklab/sdcflows/pull/30#discussion_r346640162),
although the re-centering did not seem to make a big difference.

This implementation of the rescaling uses percentiles to robustly
calculate the range of the input image (as recommended in the original
paper about FSL PRELUDE), and makes sure the output data type is
adequate (float32).

effigies commented on sdcflows/interfaces/fmap.py:
```diff
@@ -531,3 +552,22 @@ def _delta_te(in_values, te1=None, te2=None):
+def au2rads(in_file, newpath=None):
+    """Convert the input phase difference map in arbitrary units (a.u.) to rads."""
+    im = nb.load(in_file)
+    data = im.get_fdata().astype('float32')
```
Not a big deal for a single volume, but this can avoid casting to float64 before casting to float32.

```suggestion
    data = im.get_fdata(dtype='float32')
```

effigies commented on sdcflows/interfaces/fmap.py:
```diff
@@ -531,3 +552,22 @@ def _delta_te(in_values, te1=None, te2=None):
+    data[data > 2.0 * np.pi] = 2.0 * np.pi
+    hdr.set_data_dtype(np.float32)
+    hdr.set_xyzt_units('mm')
+    hdr['datatype'] = 16
```
This is a duplicate of `hdr.set_data_dtype(np.float32)` for NIfTI-1/2.

oesteban committed:
Update sdcflows/interfaces/fmap.py [skip ci]

Co-Authored-By: Chris Markiewicz <effigies@gmail.com>

oesteban committed:
enh: remove superfluous data type setting

oesteban committed:
fix: add recenter node, improve documentation

oesteban committed:
docs: fix missing reference

oesteban committed:
fix: apply recenter only within the mask

FSL PRELUDE will return some values unwrapped below zero, making it hard
to differenciate these pixels from the mask.
This commit addresses that particular issue.

oesteban committed:
fix: implement a bilateral scaling around the mode, better recentering

oesteban committed:
fix: address https://github.com/poldracklab/sdcflows/pull/50#issuecomment-555318605

effigies commented on sdcflows/interfaces/fmap.py:
```diff
@@ -531,3 +561,32 @@ def _delta_te(in_values, te1=None, te2=None):
+    """Convert the input phase difference map in arbitrary units (a.u.) to rads."""
+    from scipy.stats import mode
+    im = nb.load(in_file)
+    data = im.get_fdata(dtype='float32')
```
Now I'm thinking about this again, I wonder if it makes more sense to make it float64 for the calculation, given that it's only one volume. This would ensure that the rounding error of casting to float32 on write (the `set_data_dtype` below will still take care of that) dominates the arithmetic rounding errors.

oesteban replied:
Feels like splitting hairs... especially for images that originally were int16, DYT?

effigies replied:
Yeah, probably. But it's cheap and reduces accumulated error. That accumulated error is almost certainly in the noise, but the extra guard bits also give you lots of room to have suboptimal order of operations without changing the output relative to an optimal (smallest induced error) algorithm, which feels useful for reproducibility/comparability with other tools.

oesteban commented on sdcflows/interfaces/fmap.py:
```diff
@@ -187,7 +187,16 @@ class _Phasediff2FieldmapOutputSpec(TraitedSpec):
+    This interface is equivalent to running the following steps:
+      #. Convert from rad to rad/s
+         (``niflow.nipype1.workflows.dmri.fsl.utils.rads2radsec``)
+      #. FUGUE execution: fsl.FUGUE(save_fmap=True)
```
```suggestion
      #. FUGUE execution: fsl.FUGUE(save_fmap=True).
```
